File: backend/services/signing_summary.py
"""One row of the officer's signing list, with its key set enforced.

═══ WHY THIS MODULE EXISTS ═══

`GET /signing-requests/v2` built its rows inline in the router, and TWO
screens declared what those rows contain: the agenda at `/signings`
declared fourteen fields and the merged tracker declared eleven. Nothing
compared the three declarations, so nothing objected.

The three the tracker was missing were `live`, `days_waiting` and
`stale` — exactly the fields CANCEL1 and DASH1 added so that a screen
would stop deciding for itself which signings are over and which have
gone quiet. A screen that does not declare a field cannot render it, so
the merged tracker listed cancelled and expired signings among the live
ones and could not mark a single stuck request. The signal the agenda's
entire design leads with was absent, and no test failed.

That is FLOW1 item 0's defect one endpoint over, and milder only by
luck: there the key names were WRONG and rendered as `undefined`; here
they were ABSENT and rendered as nothing at all. Both are one contract
declared in two languages with no referee between them.

═══ THE ASSERTION IS NOT A FILTER ═══

Same reasoning as `shared_deed_row`: the key set is asserted equal to
the corpus before the row leaves, rather than being filtered down to it.
Silently dropping an unexpected key is how the screen and the server
come to disagree again, quietly — which is the failure mode this whole
pattern exists to end.

Note what this module deliberately does NOT do. It computes no state,
decides nothing about staleness, and composes no sentence: the caller
hands in what `signing_loop` and `officer_queue` already decided. A
second opinion about which states are over is the defect, not the fix.
"""

import json
from pathlib import Path
from typing import Any, Dict, Mapping, Optional

_CORPUS = json.loads(
    (Path(__file__).parent / "signing_summary_keys.json").read_text(encoding="utf-8")
)
SIGNING_SUMMARY_KEYS = frozenset(_CORPUS["keys"])


def _iso(value: Any) -> Optional[str]:
    """Timestamps cross the wire as ISO-8601 or as nothing.

    NOT as `""`. `new Date("")` is a Date object that is merely invalid,
    so every JS guard shaped `if (d)` waves it through and the screen
    renders "Invalid Date". `null` fails that guard honestly.
    """
    if value is None:
        return None
    if hasattr(value, "isoformat"):
        return value.isoformat()
    return str(value)


def signing_summary_row(
    row: Mapping[str, Any],
    *,
    state: str,
    live: bool,
    summary: str,
    days_waiting: Optional[int],
    stale: bool,
    signers: int,
) -> Dict[str, Any]:
    """One signing, as every officer surface receives it.

    The judgements arrive as arguments because they are made elsewhere
    and must keep being made elsewhere — `state`, `live` and `summary`
    are `services/signing_loop`'s, `days_waiting` and `stale` are
    `services/officer_queue`'s. This function's only job is the shape.
    """
    out: Dict[str, Any] = {
        "id": row.get("id"),
        "deed_id": row.get("deed_id"),
        "property_address": row.get("property_address"),
        "deed_type": row.get("deed_type"),
        "notary_name": row.get("notary_name"),
        "state": state,
        # CANCEL1 item 4: whether this one is still somebody's problem,
        # decided in Python. The alternative — a screen listing which
        # states are over — is that judgement copied into TypeScript,
        # where it gets missed the day a state is added.
        "live": live,
        # The server's sentence, rendered verbatim by every surface
        # (§13 rule 3). No screen composes its own account of a
        # scheduling state.
        "summary": summary,
        "booked_at": _iso(row.get("booked_at")),
        "booked_by": row.get("booked_by"),
        # FLOW1 item 4: WHEN SHE ASKED. Without it the agenda
        # reconstructed a request's age as `expires_at minus 21 days` —
        # default_expiry()'s constant, retyped as a magic number in
        # another language, so changing the default would have silently
        # re-aimed every stuck badge.
        "created_at": _iso(row.get("created_at")),
        # DASH1: how long, and whether that is too long. One threshold,
        # in services/officer_queue.py, and the screens render what they
        # are told.
        "days_waiting": days_waiting,
        "stale": stale,
        "expires_at": _iso(row.get("expires_at")),
        "signers": signers,
    }
    assert set(out) == SIGNING_SUMMARY_KEYS, (
        "the signing summary row no longer matches its contract: "
        f"extra={sorted(set(out) - SIGNING_SUMMARY_KEYS)} "
        f"missing={sorted(SIGNING_SUMMARY_KEYS - set(out))}"
    )
    return out

File: frontend/src/features/signing/signingSummary.ts
/**
 * One signing, as every officer surface receives it — and how they group.
 *
 * ═══ WHY THE TYPE LIVES HERE AND NOT ON A PAGE ═══
 *
 * `GET /signing-requests/v2` had TWO screens declaring what its rows
 * contain: the agenda declared fourteen fields, the merged tracker
 * declared eleven, and nothing compared them.
 *
 * The three the tracker was missing were `live`, `days_waiting` and
 * `stale` — the fields CANCEL1 and DASH1 added so that a screen would
 * stop deciding for itself which signings are over and which have gone
 * quiet. A screen that does not declare a field cannot render it, so the
 * merged tracker listed cancelled and expired signings among the live
 * ones and could not mark a single stuck request. No test failed.
 *
 * That is FLOW1 item 0's defect one endpoint over, and milder only by
 * luck: there the names were WRONG and rendered as `undefined`; here
 * they were ABSENT and rendered as nothing at all.
 *
 * One declaration, checked against `backend/services/signing_summary_keys.json`
 * — the same corpus the Python row builder asserts its output against, so
 * drift costs two deliberate edits instead of one silent omission.
 *
 * ═══ AND WHY THE GROUPING IS A FUNCTION ═══
 *
 * Booked / being arranged / closed was a `useMemo` on a page. As a page
 * expression its only available pin is "the source contains a filter" —
 * a string-presence pin over a decision, which cannot tell REACHABLE
 * from PRESENT. Called, it can be asked what it does with a cancelled
 * booked signing, which is the case where two plausible rules disagree.
 *
 * Note what is NOT decided here: WHICH STATES COUNT AS OVER. That is
 * `services/signing_loop.is_live`, the payload carries the verdict, and
 * this module reads `row.live` rather than holding a list. A list here
 * is the copy that gets missed the day a seventh state is added.
 */

export interface SigningSummary {
  id: number;
  deed_id: number;
  property_address: string | null;
  deed_type: string | null;
  notary_name: string | null;
  state: string;
  /** Whether this one is still somebody's problem. The SERVER's verdict. */
  live: boolean;
  /** The server's sentence about this request. Rendered verbatim (§13 rule 3). */
  summary: string;
  booked_at: string | null;
  booked_by: string | null;
  created_at: string | null;
  /** How long she has waited. The server counts; this screen does not. */
  days_waiting: number | null;
  /** Whether that is too long. The server judges; this screen does not. */
  stale: boolean;
  expires_at: string | null;
  signers: number;
}

export const STATE_LABEL: Record<string, string> = {
  requested: 'Waiting on the notary',
  windows_posted: 'Waiting on signers',
  partially_agreed: 'Part-agreed',
  booked: 'Booked',
  cancelled: 'Cancelled',
  expired: 'Expired',
};

/**
 * DASH1 — `STUCK_AFTER_DAYS = 5` used to live on the agenda page.
 *
 * The dashboard needed the same judgement, and a second threshold in
 * Python beside one in TypeScript is exactly how the partner category
 * list came to have four divergent copies. The number lives in
 * `backend/services/officer_queue.py`; the server sends `stale`; this
 * reads it.
 */
export function isStuck(row: SigningSummary): boolean {
  return row.stale;
}

export interface SigningGroups {
  booked: SigningSummary[];
  arranging: SigningSummary[];
  closed: SigningSummary[];
}

/**
 * The three groups, each sorted by the fact it actually has.
 *
 * ═══ WHY THREE AND NOT ONE SORTED LIST ═══
 *
 * The server orders by `COALESCE(booked_at, expires_at)`. For a booked
 * signing that is the signing's time; for an unbooked one it is WHEN THE
 * LINK DIES — a fact with no relationship to when anybody will meet. Two
 * orthogonal facts sharing one sort key, under a subtitle describing the
 * result as a schedule.
 *
 * So: booked by when, because that is a date. Being arranged by how long
 * she has been waiting, oldest first, because the oldest is the one that
 * needs a phone call and it has no date to be sorted by. Closed last and
 * never deleted — a cancelled request that HAD a booked time still had
 * one (T-5), and folding it away would make that unaskable.
 */
export function groupSignings(rows: SigningSummary[]): SigningGroups {
  return {
    /**
     * NO `live &&` GUARD HERE, AND THE REASON IS WORTH WRITING DOWN.
     *
     * A cancelled signing that had been booked looks like it should
     * appear in two groups — `state === 'booked'` and `!live` — and a
     * first draft of this function guarded against exactly that.
     *
     * It cannot happen, because `signing_loop.request_state` tests
     * `cancelled_at` BEFORE `booked_at`: a cancelled request reports
     * `cancelled`, never `booked`, whatever was arranged. The guard was
     * dead code carrying a false explanation, which is worse than no
     * guard — the next reader would have believed the collision was real
     * and defended against it somewhere else too.
     *
     * That ordering is now load-bearing for this grouping, so it is
     * pinned where it lives rather than assumed here:
     * `test_cancellation_beats_booking_in_the_state_vocabulary`.
     */
    booked: rows
      .filter((r) => r.state === 'booked')
      .sort((a, b) => (a.booked_at || '').localeCompare(b.booked_at || '')),
    arranging: rows
      .filter((r) => r.live && r.state !== 'booked')
      .sort((a, b) => (a.created_at || '').localeCompare(b.created_at || '')),
    closed: rows.filter((r) => !r.live),
  };
}
